_posts/drafts/hack-go-type-system.md

```diff
@@ -0,0 +1,842 @@
+---
+published: false
+title: Hacking Go's type system
+layout: post
+---
+
+Are you in the mood for a stroll inside Go's type system ?
+If you are alredy familiarized with it, this post can be funny
+for you, or just plain stupid.
+
+If you have no idea how types and interfaces are implemented on Go,
+you may learn something, I sure did :-)
+
+<!-- more -->
+
+Since I worked with handwritten type systems in C,
+like the one found in [glib GObjects](https://developer.gnome.org/gobject/stable/),
+I'm always curious on how languages implement the
+concept of type safety on a machine that actually only
+has numbers. This curiosity has extended to how I could
+bend Go's type system to my own will.
+
+Go's instances do not carry type information on the instances,
+so my only chance will involve using a interface{}. All type systems
+that I'm aware off usually implement types as some sort of integer code,
+which is used to check if the type corresponds to the one you are casting.
+
+To begin the exploration I wanted to find how type assertions are
+made, so I wrote this ridiculous code:
+
+```
+package main
+
+import "fmt"
+
+func main() {
+	var a int
+	var b interface{} = a
+	c := b.(int)
+	fmt.Println(c)
+}
+```
+
+Compiled it outputing it's assembly:
+
+```
+go build -gcflags -S cast.go
+```
+
+Found that the assembly code corresponding to this:
+
+```
+	c := b.(int)
+```
+
+Is this:
+
+```
+	0x0075 00117 (cast.go:8)	MOVQ	CX, 8(SP)
+	0x007a 00122 (cast.go:8)	MOVQ	AX, 16(SP)
+	0x007f 00127 (cast.go:8)	LEAQ	"".autotmp_1+64(SP), AX
+	0x0084 00132 (cast.go:8)	MOVQ	AX, 24(SP)
+	0x0089 00137 (cast.go:8)	PCDATA	$0, $0
+	0x0089 00137 (cast.go:8)	CALL	runtime.assertE2T(SB)
+```
+
+The **runtime.assertE2T** call caught my attention, it was not hard to
+find it on the [iface.go](https://github.com/golang/go/blob/master/src/runtime/iface.go)
+file on the golang source code.
+
+It's code (on the time of writing, Go 1.7.3):
+
+```
+func assertE2T(t *_type, e eface, r unsafe.Pointer) {
+	if e._type == nil {
+		panic(&TypeAssertionError{"", "", t.string(), ""})
+	}
+	if e._type != t {
+		panic(&TypeAssertionError{"", e._type.string(), t.string(), ""})
+	}
+	if r != nil {
+		if isDirectIface(t) {
+			writebarrierptr((*uintptr)(r), uintptr(e.data))
+		} else {
+			typedmemmove(t, r, e.data)
+		}
+	}
+}
+```
+
+There is the eface type, that has a **_type** field. Checking out what
+would be a **eface** I found this:
+
+```go
+type iface struct {
+	tab  *itab
+	data unsafe.Pointer
+}
+
+type eface struct {
+	_type *_type
+	data  unsafe.Pointer
+}
+```
+
+From what I read before on
+[Russ Cox post about interfaces](https://research.swtch.com/interfaces)
+I would guess that the **iface** is used when you are using
+interfaces that have actually methods on it (that is why it has
+a itab, the interface table, which is roughly equivalent to
+a C++ vtable).
```

Comment:
I don't understand why I'm so addicted to parenthesis =/...Lisp calling perhaps ?